warehouse/router/sync_test.go

```diff
@@ -0,0 +1,261 @@
+package router
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	miniogo "github.com/minio/minio-go/v7"
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/require"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource/minio"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource/postgres"
+
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+	sqlmiddleware "github.com/rudderlabs/rudder-server/warehouse/integrations/middleware/sqlquerywrapper"
+	"github.com/rudderlabs/rudder-server/warehouse/internal/model"
+	"github.com/rudderlabs/rudder-server/warehouse/schema"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+type mockFetchSchemaRepo struct {
+	schemaInWarehouse model.Schema
+	err               error
+}
+
+func (m *mockFetchSchemaRepo) FetchSchema(_ context.Context) (model.Schema, error) {
+	if m.err != nil {
+		return model.Schema{}, m.err
+	}
+	return m.schemaInWarehouse, nil
+}
+
+type mockSyncSchemaRepo struct {
+	getLocalSchemaerr    error
+	updateLocalSchemaErr error
+	fetchSchemaErr       error
+	hasChanged           bool
+	schemaMap            map[string]model.Schema
+}
+
+func schemaKey(sourceID, destinationID, namespace string) string {
+	return fmt.Sprintf("%s_%s_%s", sourceID, destinationID, namespace)
+}
+
+func (m *mockSyncSchemaRepo) GetLocalSchema(_ context.Context) (model.Schema, error) {
+	if m.getLocalSchemaerr != nil {
+		return model.Schema{}, m.getLocalSchemaerr
+	}
+
+	return model.Schema{}, nil
+}
+
+func (m *mockSyncSchemaRepo) UpdateLocalSchemaWithWarehouse(_ context.Context, _ model.Schema) error {
+	if m.updateLocalSchemaErr != nil && m.hasChanged {
+		return m.updateLocalSchemaErr
+	}
+	return nil
+}
+
+func (m *mockSyncSchemaRepo) HasSchemaChanged(_ model.Schema) bool {
+	return m.hasChanged
+}
+
+func (m *mockSyncSchemaRepo) FetchSchemaFromWarehouse(_ context.Context, _ schema.FetchSchemaRepo) (model.Schema, error) {
+	if m.fetchSchemaErr != nil {
+		return nil, m.fetchSchemaErr
+	}
+
+	return m.schemaMap[schemaKey("test-sourceID", "test-destinationID", "test-namespace")], nil
+}
+
+func TestSync_SyncRemoteSchema(t *testing.T) {
+	t.Run("fetching schema from local fails", func(t *testing.T) {
+		mock := &mockSyncSchemaRepo{
+			getLocalSchemaerr: fmt.Errorf("error fetching schema from local"),
+		}
+		mockFetchSchema := &mockFetchSchemaRepo{}
+		r := &Router{
+			logger: logger.NOP,
+		}
+		err := r.syncRemoteSchema(context.Background(), mockFetchSchema, mock)
+		require.Error(t, err)
+		require.True(t, errors.Is(err, mock.getLocalSchemaerr))
```

Comment:
Same for other places as well.
```suggestion
		require.ErrorIs(t, err, mock.getLocalSchemaerr)
```